packages/legend-query-builder/src/components/watermark/QueryBuilderWatermark.tsx

```diff
@@ -0,0 +1,190 @@
+            !monitor.didDrop() &&
+            // Only allows parameters with muliplicity 1 and type string
+            item.variable.parameter.multiplicity ===
+              graph.getTypicalMultiplicity(TYPICAL_MULTIPLICITY_TYPE.ONE) &&
+            item.variable.parameter.genericType?.value.rawType.name ===
+              'String' &&
+            itemType
+          ) {
+            handleDrop(item);
+          } // prevent drop event propagation to accomondate for nested DnD
+        },
+        collect: (monitor) => ({
+          isParameterValueDragOver: monitor.isOver({
+            shallow: true,
+          }),
+        }),
+      }),
+      [handleDrop],
+    );
+
+    return (
+      <PanelFormSection>
+        <div className="query-builder__parameter-editor">
+          <PanelDropZone
+            isDragOver={isParameterValueDragOver}
+            dropTargetConnector={dropTargetConnector}
+          >
+            <BasicValueSpecificationEditor
+              valueSpecification={guaranteeNonNullable(watermarkValue)}
+              setValueSpecification={(val: ValueSpecification): void => {
+                watermarkState.setValue(val);
+              }}
+              graph={graph}
+              obseverContext={
+                watermarkState.queryBuilderState.observableContext
+              }
+              typeCheckOption={{
+                expectedType: graph.getPrimitiveType(PRIMITIVE_TYPE.STRING),
+              }}
+              resetValue={(): void => {
+                watermarkState.resetValue();
+              }}
+            />
+          </PanelDropZone>
+        </div>
+      </PanelFormSection>
+    );
+  },
+);
+
+export const QueryBuilderWatermarkEditor = observer(
+  (props: { queryBuilderState: QueryBuilderState }) => {
+    const { queryBuilderState } = props;
+    const watermarkState = queryBuilderState.watermarkState;
+    const handleClose = (): void => {
+      queryBuilderState.setIsEditingWatermark(false);
+    };
+
+    return (
+      <Dialog
+        open={Boolean(queryBuilderState.isEditingWatermark)}
+        onClose={handleClose}
+        classes={{
+          root: 'editor-modal__root-container',
+          container: 'editor-modal__container',
+          paper: 'editor-modal__content',
+        }}
+      >
+        <Modal darkMode={true} className="editor-modal">
+          <ModalHeader title="Watermark" />
+          <ModalBody>
+            <PanelForm>
+              <PanelFormBooleanField
+                isReadOnly={false}
+                value={watermarkState.value !== undefined}
+                prompt="Enable watermark"
+                update={(): void => watermarkState.enableWatermark()}
+              />
+              {watermarkState.value && (
+                <>
+                  <WatermarkValueEditor
+                    watermarkState={watermarkState}
+                    watermarkValue={watermarkState.value}
+                  />
+                  <PanelDivider />
```

Comment:
can we create a const before the rendering such as 
```
const compatibleParameters = queryBuilderState.parametersState.parameterStates.map(p => p.parameter).filter(watermarkState.isParamaterCompatibleWithWaterMark)
```
then use that when rendering 
```
   {compatibleParameters
                      .length === 0 && <> No available parameters in query </>}
```
and 
```
compatibleParameters.map(<VariableExpressionViewer...>)

```